Libraries/LibWeb/CSS/SelectorEngine.cpp

```diff
@@ -988,6 +988,55 @@ static inline bool matches_pseudo_class(CSS::Selector::SimpleSelector::PseudoCla
 
         return false;
     }
+    case CSS::PseudoClass::Required: {
+        // https://html.spec.whatwg.org/multipage/semantics-other.html#selector-required
+
+        // The :required pseudo-class must match any element falling into one of the following categories:
+        // - input elements that are required
+        if (auto const* input_element = as_if<Web::HTML::HTMLInputElement>(element)) {
+            if (input_element->required_applies() && input_element->has_attribute(HTML::AttributeNames::required))
+                return true;
+        }
+        // - select elements that have a required attribute
+        if (auto const* select_element = as_if<Web::HTML::HTMLSelectElement>(element)) {
```

Comment:
Make these `else if` perhaps? Otherwise we're doing (potentially expensive) type checks we don't need to, if eg it's an `<input>` that isn't required.